src/crimson/osd/pg.cc

```diff
@@ -1212,142 +1071,84 @@ PG::interruptible_future<std::optional<eversion_t>> PG::submit_error_log(
         get_collection_ref(), std::move(t)
       ).then([this] {
         peering_state.update_trim_to();
-        return seastar::make_ready_future<std::optional<eversion_t>>(projected_last_update);
+        return seastar::make_ready_future<eversion_t>(projected_last_update);
       });
     });
   });
 }
 
-PG::do_osd_ops_iertr::future<PG::pg_rep_op_fut_t<MURef<MOSDOpReply>>>
-PG::do_osd_ops(
-  Ref<MOSDOp> m,
-  crimson::net::ConnectionXcoreRef conn,
+PG::run_executer_fut PG::run_executer(
+  OpsExecuter &ox,
   ObjectContextRef obc,
   const OpInfo &op_info,
-  const SnapContext& snapc)
+  std::vector<OSDOp>& ops)
 {
-  if (__builtin_expect(stopping, false)) {
-    throw crimson::common::system_shutdown_exception();
+  LOG_PREFIX(PG::run_executer);
+  auto rollback_on_error = seastar::defer(
+    [rollbacker = ox.create_rollbacker(
+	[stored_obc=duplicate_obc(obc)](auto &obc) mutable {
+	  obc->update_from(*stored_obc);
+	})]() mutable {
+      rollbacker.rollback_obc_if_modified();
+    });
```

Comment:
The nesting lambdas seems confusing, probably there is a better way to make this more readable?

Reply:
Split out the rollbacker declaration, probably a bit better now?

Reply:
Yeah, thanks!